datapath: preserve tail call maps in between replacing interface program pairs

ti-mo opened the pull request:
To prevent unpinning tail call maps after replacing only one half of an interface's programs, e.g. `from-netdev` and `to-netdev`, allow the caller to defer finalizer operations until a batch of interfaces have been replaced.

This avoids short windows of packet drops due to missed tail calls.

Fixes #16561

Also fixed a bug where we specified the wrong bpffs path to the migration code. This causes migrations to not be executed in the currently-released version of 1.11, meaning resizing maps causes endpoints regeneration failures.

```release-note
Preserve tail call maps during resize to prevent drops during agent upgrade
```

ti-mo committed:
datapath: return finalizer from replaceDatapath

To prevent unpinning tail call maps after replacing only one half of an
interface's programs, e.g. from-netdev and to-netdev, allow the caller to
defer finalizer operations until a batch of interfaces have been replaced.

This avoids short windows of packet drops due to missed tail calls.

Fixes #16561

Signed-off-by: Timo Beckers <timo@isovalent.com>

kkourt reviewed (approved):
LGTM.

I'm not 100% sure about  not garbage collecting  the files, but if this only happens when the map size is changed (which, I guess, does not happen that often) I cannot think of a strong argument against it. 

That being said, could we do it an easy version of GC where we keep only _one_ backup directory? That is, after the backup is done we look for older backups for the same map and delete them.

pchaigno reviewed (approved):
LGTM. Thanks!

Did you already test this by adding random tail call somewhere and executing K8sUpdates?

> I'm not 100% sure about not garbage collecting the files, but if this only happens when the map size is changed (which, I guess, does not happen that often) I cannot think of a strong argument against it.

If it becomes an issue, we could maybe rely on the file creation timestamps in bpffs to determine which backups have been made a very long time ago and are safe to delete. But I really doubt Kubernetes nodes have a lifespan long enough for this to become an issue.

> That being said, could we do it an easy version of GC where we keep only one backup directory? That is, after the backup is done we look for older backups for the same map and delete them.

I don't think we can even assume datapath logic n-1 is fully loaded when we start loading datapath logic n. So there could still be bpf_lxc programs using tail call maps from the n-2 version... or even before :-/

joestringer commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -185,10 +187,50 @@ func finalizeMap(bpffsPath, name string, revert bool) error {
 		return nil
 	}
 
+	// Never unpin prog arrays. Doing so causes existing BPF progs to miss tail
```
When this says "Never", what exactly does that mean? Does that mean we will leak these maps upon each upgrade? How will they get cleaned up?

ti-mo replied:
:wave: Rationale here: https://github.com/cilium/cilium/issues/16561#issuecomment-938534720

joestringer replied:
I couldn't quite tell from the linked comment, I guess admittedly I probably would need to spend a bit more time understanding. But if the answer is really "never", then I think we need to fix this. Otherwise we are forever consuming resources in the kernel based on the number of times we restart Cilium.

ti-mo replied:
> Otherwise we are forever consuming resources in the kernel based on the number of times we restart Cilium.

As mentioned in the linked comment: `Prog array migration only happens on Cilium up/downgrade, which is not supposed to happen all too frequently`. For example, https://github.com/cilium/cilium/pull/17758 proposes to bump `CILIUM_CALL_SIZE` from 25 to 27, which will result in that tail call map being recreated with the new size, once, when the agent is upgraded. This leaves one lingering prog array with 25 max entries on the system.

The alternative is periodically scanning all progs in the kernel (which can run into the thousands on busy nodes) for references to this old map. This process might need to go on forever until all pods using the map have been killed/rotated. Scanning takes (at least) one syscall per prog in the kernel.

Not accounting for repeated up/downgrades that will create one backup map per tick, I'm not convinced that adding a GC is worth the development and runtime execution overhead given the small size of these maps. Alternatively, we could always document that these map pins can be safely deleted if all managed pods have been rotated on a node post-upgrade.

joestringer replied:
Is there a solution where we just delete this `backup` directory, thereby unpinning after some time? Or track `backup1` and `backup2` and then upon upgrade, we always move the current ones to `backup1`, move the old `backup1` to `backup2` and make sure the old `backup2` are deleted?

ti-mo replied:
> Is there a solution where we just delete this `backup` directory, thereby unpinning after some time?

Note that deleting an in-use tail call map results in dropped packets. Establishing the `after some time` heuristic is the problem here. For example, interrupted connectivity 10 mins after an upgrade would be even more surprising than interruptions during an upgrade.

> Or track backup1 and backup2 and then upon upgrade, we always move the current ones to backup1, move the old backup1 to backup2 and make sure the old backup2 are deleted?

Not sure how we would track whether `backup1` is current, old or new without keeping persistent state on the host. It's not clear what exactly you're proposing here. As Paul mentioned above, it's not even guaranteed that all pods created using Cilium version n-2 have died before the system was upgraded to n-1. There is always a risk of interruptions when blindly (without refscanning) unpinning a tail call map.

joestringer replied:
:+1: indeed I agree that "after some fixed time" is not a reasonable solution.

One thing I don't understand is why we don't know if the map is still in use or not. In general, upon startup Cilium is intended to replace all running eBPF programs (previously installed by earlier Cilium) with new programs to ensure that the datapath is now running the new datapath implementation code. We should know exactly all of the possible points that should be configured and attached. Broadly this consists of the core programs eg attached to netdevices, sockets etc. plus per-endpoint programs. Then if there is anything else left over that we didn't migrate to the new version, we should be able to safely remove it. We already persist the information necessary to determine this.

pchaigno replied:
> One thing I don't understand is why we don't know if the map is still in use or not.

We can know that, but we'd have to parse the xlated BPF programs from the kernel to search for all references to the tail call map.

> +1 indeed I agree that "after some fixed time" is not a reasonable solution.

It is not because you're both talking minutes. Are your concerns with that approach still valid if we delete maps older than a week?

joestringer replied:
> We can know that, but we'd have to parse the xlated BPF programs from the kernel to search for all references to the tail call map.

Where I'm coming from is, we have a set of known intended state in the agent, and we will then conform all the hook points that we control to the intended state. This approach says "I don't want to analyse the current situation from the kernel perspective and handle all possible references that exist", it says "If there are known paths on the filesystem for pinned programs, we will move those out of the way and then instantiate the new programs in the old places. When I am done, everything should be in the correct state and there should be no remaining references to the old programs.". If we are still missing cases there, then we aren't doing enough up-front tracking of the intended state of the system.

> It is not because you're both talking minutes. Are your concerns with that approach still valid if we delete maps older than a week?

I think ultimately this is worse. Rather than "hmm, connectivity for my pod breaks after N minutes following upgrade", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not. I'd much rather do the wrong thing early with some possibility of understanding the trigger rather than doing the wrong thing much later and have no idea why it occurred. (Of course, I'd rather do the right thing but I'm covering worst-case scenarios here :sweat_smile: )

pchaigno replied:
> Rather than "hmm, connectivity for my pod breaks after N minutes following upgrade", we receive reports of "connectivity randomly broke" and then when we ask "Did you upgrade recently?" and they say "well maybe a week ago" then we will really scratch our head about whether the failure is related to upgrade or not.

Uh, wouldn't this only happen if the endpoints fail to be regenerated for over a week? I think we have more serious issues if that happens :grimacing:

joestringer replied:
Is that true? That feels like it frames the resource cleanup question much more nicely.

If everything succeeds regeneration, delete the `backup` directory. Otherwise we need to keep it around to keep the endpoints successfully connected, and hopefully the warning signs from regeneration will reach the user and they'll roll back to a known good state, ie previous version. In that particular case there would still be a `backup` directory on the system until they next upgrade, but at that point we could just remove/overwrite those `backup` pins and hopefully have better luck the second time (presumably with a new release of the target upgrade version).

joestringer commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -185,10 +187,50 @@ func finalizeMap(bpffsPath, name string, revert bool) error {
+	// references to determine if it's safe to delete.
+	if pending.Type() == ebpf.ProgramArray {
+		log.WithFields(logrus.Fields{logfields.BPFMapPath: file, logfields.BPFMapName: name}).
+			Info("Backing up program array after successful recreation")
```
Do we really need to log this at info level? Seems like a very specific implementation detail which means nothing to users.

ti-mo replied:
This feels like a crucial event to have in the logs as it's only performed during upgrades when map properties have changed. Especially given the troubleshooting history that led to this PR, being able to observe this event (or a lack thereof) could be important in some situations.

I'll leave this up to @brb or @borkmann to decide, but will leave it in for now.

brb replied:
I'd keep the log statement at least until we have enough confidence that the new map migrate infrastructure is sound.

joestringer commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -185,10 +187,50 @@ func finalizeMap(bpffsPath, name string, revert bool) error {
+	// Ensure the backup dir exists.
+	backupDir := filepath.Join(bpffsPath, "backup")
+	if err := os.MkdirAll(backupDir, os.ModePerm); err != nil {
+		return fmt.Errorf("unable to create bpffs backup directory: %v", err)
```
```suggestion
		return fmt.Errorf("unable to create bpffs backup directory: %w", err)
```

ti-mo committed:
bpf,datapath: map-migrate - specify full tc/globals/ bpffs path

This patch fixes bpffs migrations meant to occur at runtime. The tc/globals/
part of the path was missing, and the bpffs root was being scanned instead,
causing migrations not to run. This would lead to endpoints continuously
failing to regenerate at runtime when map definitions in the ELF differ from
their pinned counterparts, e.g. during an agent upgrade.

For example, when resizing CALLS_MAP and trying to replace a program, the old
map will be loaded from bpffs, while the program expects to see a map of a
different capacity:

```
~ tc filter replace dev cilium-probe ingress bpf da obj bpf/bpf_lxc.o sec from-container
libbpf: couldn't reuse pinned map at '/sys/fs/bpf/tc//globals/test_cilium_calls_65535': parameter mismatch
libbpf: map 'test_cilium_calls_65535': error reusing pinned map
libbpf: map 'test_cilium_calls_65535': failed to create: Invalid argument(-22)
libbpf: failed to load object 'bpf/bpf_lxc.o'
Unable to load program
```

The CLI equivalent executed as part of init.sh does work as expected, as the
full path including tc/globals/ is specified there.

Signed-off-by: Timo Beckers <timo@isovalent.com>

pchaigno commented on pkg/datapath/loader/netlink.go:
```diff
@@ -70,7 +70,7 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
 	// Temporarily rename bpffs pins of maps whose definitions have changed in
 	// a new version of a datapath ELF.
-	if err := bpf.StartBPFFSMigration(bpf.GetMapRoot(), objPath); err != nil {
+	if err := bpf.StartBPFFSMigration(bpf.MapPrefixPath(), objPath); err != nil {
```
Nice catch!

Maybe one explanation why we didn't get more bug reports from this error is because we may not have changed that many maps between v1.10 and v1.11. And if we did, it may have been only for advanced features, disabled by default (so not covered in upgrade test).

pchaigno commented on pkg/datapath/loader/netlink.go:
```diff
@@ -70,7 +70,7 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
 	// Temporarily rename bpffs pins of maps whose definitions have changed in
 	// a new version of a datapath ELF.
-	if err := bpf.StartBPFFSMigration(bpf.GetMapRoot(), objPath); err != nil {
+	if err := bpf.StartBPFFSMigration(bpf.MapPrefixPath(), objPath); err != nil {
```
Are we confident enough about the second fix to backport it to v1.11 straight away? That would mean it will be included in v1.11.2, which larger production users are likely to upgrade to. If not, then I think we'll need to move the first bugfix to a separate PR, to be able to backport it now without backporting the more impactful change from the second commit.

cc @joestringer for a maintainer's perspective.

ti-mo replied:
I'm open to carving out the first fix (bpffs path) into a separate PR. I think it's rather critical given it can deadlock datapath regeneration (or even prevent agent start?) when maps get changed.

The second (deferred map removal) is mostly oriented towards 1.12, but I was considering the case where we need to add and backport a tail call as an 'emergency' if a complexity issue arises that we cannot solve any other way. It's completely up to you to; perhaps we should only backport this fix if/when such a situation actually arises.

ti-mo replied:
On second thought, I think this is good to go. I don't see this as a particularly invasive or risky change, it just defers cleanup to be slightly later.

joestringer replied:
>  I think it's rather critical given it can deadlock datapath regeneration (or even prevent agent start?) when maps get changed.

Can you describe this in a bit more detail, also include it in the commit message? It's not immediately obvious to me which conditions would cause this kind of situation or even that the first patch is resolving such a serious bug.

> On second thought, I think this is good to go. I don't see this as a particularly invasive or risky change, it just defers cleanup to be slightly later.

:heavy_plus_sign: Assuming we resolve the other feedback, this should be fairly low risk.

pchaigno commented on pkg/datapath/loader/loader.go:
```diff
@@ -315,9 +319,11 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 			}
 			return err
 		}
+		finalize()
```
I thought the goal of this callback function was to only remove maps once all BPF programs for the interface have been replaced. But if `ep.RequireEgressProg()` is true below, then we haven't yet replaced the egress program for the interface when we execute this `finalize`.

ti-mo replied:
Good catch. I'll defer both finalizers here.

pchaigno reviewed (commented):
Nice catch on both bugs! In particular, the second fix is definitely simpler than other solutions we had in mind. Not sure how we missed that solution before :facepalm: 

(If we run into any issue later on with the second bugfix, I think another approach could be to simply separate the per-endpoint tail call map into ingress and egress tail call maps. AFAIK, it doesn't actually need to be shared between both directions.)

My one concern with this is on how we want to backport, as outlined below. Are we confident enough in the second bugfix to backport it right away? Did you perform any upgrade tests other than the CI's? How did you check that it is fixing #16561?

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -98,25 +96,36 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
+		}
+	}
+
+	return finalize, nil
```
The comment on line 99 states that if the load fails, the original maps will be repinned to their original paths. However, in that case (lines 104-106) the code only sets `revert = true` and then returns a `nil` finalizer. As a result, `bpf.FinalizeBPFFSMigration()` does not get called.

In the successful case, I think that we need to defer the map migration finalization until after all of the programs have been loaded, so that case makes sense to pass the finalizer back to the caller.

In the error case where the program cannot be loaded, I think it is OK to just run the contents of this finalizer function directly from here with `revert=true` and pass back a nil finalizer (like the code on line 106 does already).

ti-mo replied:
Oops, good catch. I opted to get rid of the bool and call finalize directly when handling the error. I also specified this behaviour in the `replaceDatapath` docstring.

joestringer commented on pkg/datapath/loader/loader.go:
```diff
@@ -315,9 +319,11 @@ func (l *Loader) reloadDatapath(ctx context.Context, ep datapath.Endpoint, dirs
 
 		if ep.RequireEgressProg() {
-			if err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, false, ""); err != nil {
+			finalize, err := replaceDatapath(ctx, ep.InterfaceName(), objPath, symbolToEndpoint, dirEgress, false, "")
```
If I follow right, the approach here is to ensure that both the ingress+egress programs for a given endpoint are completely reloaded before unpinning the old references. That makes some sense to me.

Similarly for the `reloadHostDatapath` case above, the programs for all interfaces are loaded first before unpinning the old map references. That also makes sense. :+1: 

I recall that some of the discussion previously was around what happens when we reload base programs, and how that could impact upgrade. Looking through, it seems like we've managed to sidestep that question for now while resolving the immediate concerns around upgrade. Is that right? The approach here seems like a good incremental improvement and allows us to defer dealing with the `init.sh` conversion and this other potential datapath disruption question until the go migration work matures.

(no changes necessary, this is more just me checking if I am following correctly)

ti-mo replied:
Yup, I think we're on the same page here. Could you specify what you mean by 'base programs'?

joestringer replied:
Sure thing. Base programs are the "global" programs that apply once for the node, like bpf_host, bpf_overlay, etc. Structurally in the code, when Cilium wants to update these programs, it calls into the datapath [Reinitialize()](https://github.com/cilium/cilium/blob/477cccaa2afad1aaa49e29dec80952fc050844c7/pkg/datapath/loader/base.go#L237) function. It's typically pretty rare to reinitialize the datapath like this during normal Cilium operations.

This is in contrast to the per-endpoint `bpf_lxc` programs which are cloned per-endpoint and are dynamically configured based on creation of workloads on the nodes. These endpoint BPF programs are installed via the endpoint-specific code in `pkg/endpoint/bpf.go`. Thes actual program reload for endpoints goes via either:
* `CompileAndLoad()`: unusual; typically only triggered by `cilium endpoint regenerate` CLI,
* [CompileOrLoad()](https://github.com/cilium/cilium/blob/477cccaa2afad1aaa49e29dec80952fc050844c7/pkg/endpoint/bpf.go#L699): typical case on startup, or
* `ReloadDatapath()`: used when updating maps for individual endpoints, does not trigger recompilation of ELFs and does not reload the global programs.

ti-mo replied:
@joestringer Thanks, I thought these could only regenerate on agent startup, TIL `endpoint regenerate`.

joestringer commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -119,6 +121,10 @@ func parseExtra(spec *ebpf.MapSpec, coll *ebpf.CollectionSpec) error {
 func repinMap(bpffsPath string, name string, spec *ebpf.MapSpec) error {
+	if _, err := os.Stat(bpffsPath); errors.Is(err, fs.ErrNotExist) {
+		return fmt.Errorf("missing bpffs dir: %w", err)
+	}
```
nit, Maybe a dumb question here, but the commit message says:

> The `tc/globals/` part of the path was missing, and the bpffs root was being scanned instead. Prevent this error from going unnoticed in the future by explicitly checking if the map directory exists.

If the directory path being passed in was `/sys/fs/bpf` instead of `/sys/fs/bpf/tc/globals`, then won't the directory check pass regardless of the bug? Also if we're part way through migrating these maps, then presumably the directories already exist somewhere, so why is it helpful to make these error checks here?

If there's a reasonable reason, then I'm fine with keeping these, my above questions are really just assessing what kind of difference they make and whether we need these at all (here and in `finalizeMap()`).

ti-mo replied:
I see this more as a courtesy to the user/caller, e.g. when `/sys/fs/bpf` is not mounted or `tc/globals` not present for some reason.

Actually, now I think of it, this might do more harm than good if migrations are run before iproute2 had the chance to create `tc/globals`. Missing pins are ignored (system already in desired state), but missing `tc/globals` is an error. I'll remove the checks, they don't really belong there.

joestringer commented on pkg/datapath/loader/netlink.go:
```diff
@@ -98,14 +99,26 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
 
-	return nil
+	finalize := func() {
+		l := log.WithField("device", ifName).WithField("objPath", objPath)
```
FYI there's also `log.WithFields(logrus.Fields{ ... })` to specify multiple fields at once with one call. This is a more common style elsewhere in the codebase, but also this doesn't make any meaningful difference so don't worry about changing this one.

joestringer reviewed (approved):
LGTM, thanks again for the fix.
